muselog/context.py

```diff
@@ -28,23 +29,49 @@ def clear() -> None:
     ctx.clear()
 
 
-def bind(**ctx) -> None:
-    """Put keys and values into the context-local context."""
+def bind(**ctx):
+    """Put keys and values into the context-local context.
+
+    NOTE: Keys in the ctx parameter overwrite keys of the same name in the
+    global context.
+    """
     _CONTEXT.get().update(ctx)
 
 
-def unbind(*keys) -> None:
-    """Remove *keys* from the context-local context if they are present."""
+def unbind(*keys):
+    """Remove *keys* from the context-local context, if they are present."""
     ctx = _CONTEXT.get()
     for key in keys:
         ctx.pop(key, None)
 
 
-@contextmanager
-def Context(**ctx) -> Iterator:
-    """Bind additional context to muselog global context."""
-    bind(**ctx)
-    try:
-        yield
-    finally:
-        unbind(*list(ctx.keys()))
+class Context:
+    """Scoped context management.
+
+    This class provides a context manager that will automatically
+    remove only that context which it was provided.
+
+    It also provides a `bind` method that allows adding to the
+    context, and removing those newly added items on context
+    manager exit.
+
+    Note that Context will not work as expected when used with generators.
+    See https://www.python.org/dev/peps/pep-0568/ for more information.
+    """
+
+    def __init__(self, **ctx):
+        """Initialize the context with the given keyword arguments."""
+        self.ctx = ctx
+
+    def __enter__(self) -> Context:  # noqa: D105
+        self.bind()
```

Comment:
Why do we bind on enter if there is no context provided to bind?

Reply:
Ah never mind, it's so that we can pass in a context on init that gets bound on enter